installer/arch/bootstrap

```diff
@@ -0,0 +1,150 @@
+    tar tf "$FETCHDIRDB/$REPO.db" \
+        | sed -n 's|^\(.*\)/$|'"$REPO:"'\1|p' >> "$LIST"
+done
+
+# Get the value of a field in the Arch package database
+# getvaluedb repo pkgver [desc|depends] field
+getdbfield() {
+    REPO="$FETCHDIRDB/$1.db"
+    FILE="$2/$3"
+    FIELD="$4"
+    # Extract the relevant file
+    tar xf "$REPO" -O "$FILE" | awk '
+        /^$/ { m = 0 }
+        m { print }
+        /^%'"$FIELD"'%$/ { m = 1 }
+    '
+}
+
+echo "Downloading and extracting packages..." 1>&2
+missing="$PACKAGES_BOOTSTRAP"
+installed=""
+bootstrappkg=""
+
+# Install packages, taking care of dependencies:
+# - Install all packages in $missing that have not been installed already.
+# - Record all dependencies in $nextmissing
+# - Set missing=nextmissing
+# - Loop until $missing is empty
+
+while [ -n "$missing" ]; do
+    nextmissing=""
+    for PACKAGE in $missing; do
+        # Do not install if already installed
+        for IPKG in $installed; do
+            if [ "$IPKG" = "$PACKAGE" ]; then
+                continue 2
+            fi
+        done
+
+        echo "Downloading and installing $PACKAGE..." 1>&2
+        PKG="`grep ":$PACKAGE-[0-9]" "$LIST" | head -n 1`"
+        if [ -z "$PKG" ]; then
+            echo "Cannot find package $PACKAGE..." 1>&2
+            exit 1
+        fi
+        REPO="`echo "$PKG" | cut -f 1 -d:`"
+        PKGVER="`echo "$PKG" | cut -f 2 -d:`"
+        FILE="`getdbfield "$REPO" "$PKGVER" "desc" "FILENAME"`"
+        MIRRORBASE="`echo "$MIRROR" | sed -e "s/\\$repo/$REPO/" \
+                                          -e "s/\\$arch/$ARCH/"`"
+        curlretry -o "$FETCHDIRPKG/$FILE" "$MIRRORBASE/$FILE" 1>&2
+        tar --warning=no-unknown-keyword -xkf "$FETCHDIRPKG/$FILE" \
+            --exclude ".PKGINFO" --exclude ".MTREE" \
+            --exclude ".INSTALL" --exclude ".Changelog" \
+            -C "$BOOTSTRAPCHROOT"
+
+        # Get list of dependencies (ignore versions)
+        ndep="`getdbfield "$REPO" "$PKGVER" "depends" "DEPENDS" \
+                     | sed -n -e 's/^\([^<>=]*\).*$/\1/p' | tr '\n' ' '`"
+
+        # Some packages provide virtual packages (e.g. bash provides sh):
+        # Add those to the installed list
+        nprovides="`getdbfield "$REPO" "$PKGVER" "depends" "PROVIDES" \
+                          | tr '\n' ' '`"
+
+        bootstrappkg="$bootstrappkg $PACKAGE"
+        installed="$installed $nprovides $PACKAGE"
+        nextmissing="$nextmissing $ndep"
+    done
+
+    missing="$nextmissing"
+done
+
+# Move databases to bootstrap fs
+mkdir -p "$BOOTSTRAPCHROOT/var/lib/pacman/sync"
+mv "$FETCHDIRDB"/*.db "$BOOTSTRAPCHROOT/var/lib/pacman/sync"
+# Move package tarballs, so we can reinstall them properly
+mkdir -p "$BOOTSTRAPCHROOT/var/cache/pacman/pkg"
+mv "$FETCHDIRPKG"/* "$BOOTSTRAPCHROOT/var/cache/pacman/pkg"
+
+echo "arch" > "$BOOTSTRAPCHROOT/etc/hostname"
+# This is only necessary for reinstall after bootstrapping: signature check
+# is eventually enabled again.
```

Comment:
Yeah, that's tricky. pacman.conf is overwritten when the packages are reinstalled (I'll add a comment: it also matters for architecture specification).

Actually, I moved around the pacman keyring generation (before package reinstall), so that this is not necessary anymore.